reaktive/src/commonMain/kotlin/com/badoo/reaktive/observable/SwitchMap.kt

```diff
@@ -0,0 +1,78 @@
+package com.badoo.reaktive.observable
+
+import com.badoo.reaktive.base.ErrorCallback
+import com.badoo.reaktive.base.ValueCallback
+import com.badoo.reaktive.base.subscribeSafe
+import com.badoo.reaktive.base.tryCatch
+import com.badoo.reaktive.disposable.CompositeDisposable
+import com.badoo.reaktive.disposable.Disposable
+import com.badoo.reaktive.disposable.DisposableWrapper
+import com.badoo.reaktive.utils.atomic.AtomicBoolean
+import com.badoo.reaktive.utils.atomic.AtomicInt
+
+fun <T, R> Observable<T>.switchMap(mapper: (T) -> Observable<R>): Observable<R> =
+    observable { emitter ->
+        val disposables = CompositeDisposable()
+        emitter.setDisposable(disposables)
+        val serializedEmitter = emitter.serialize()
+
+        subscribeSafe(object : ObservableObserver<T>, ErrorCallback by serializedEmitter {
+            private val activeSourceCount = AtomicInt(initialValue = 1)
+            private val activeObserver = DisposableWrapper()
+
+            override fun onSubscribe(disposable: Disposable) {
+                disposables += disposable
+                disposables += activeObserver
+            }
+
+            override fun onNext(value: T) {
+                activeSourceCount.addAndGet(delta = 1)
+
+                serializedEmitter.tryCatch({ mapper(value) }) { innerObservable ->
+                    val innerObserver = object : ObservableObserver<R>, Disposable,
+                        ValueCallback<R> by serializedEmitter,
+                        ErrorCallback by this {
+
+                        private val disposableWrapper = DisposableWrapper()
+                        private val isCompleted = AtomicBoolean(initialValue = false)
+
+                        override fun onSubscribe(disposable: Disposable) {
+                            disposableWrapper.set(disposable)
```

Comment:
`onSubscribe` might be called asynchronously, so `onSubscribe` for a new inner source might be called BEFORE a previous inner source's `onSubscribe`. Please add the following test cases:
```
  @Test
    fun disposes_previous_inner_source_disposable_IF_it_is_provided_after_new_source_disposable() {
        lateinit var innerObserver1: ObservableObserver<String?>
        val inner1 = observableUnsafe<String?> { innerObserver1 = it }
        val innerDisposable1 = disposable()
        lateinit var innerObserver2: ObservableObserver<String?>
        val inner2 = observableUnsafe<String?> { innerObserver2 = it }
        switchMapUpstreamAndSubscribe(listOf(inner1, inner2))

        source.onNext(0)
        source.onNext(1)
        innerObserver2.onSubscribe(disposable())
        innerObserver1.onSubscribe(innerDisposable1)

        assertTrue(innerDisposable1.isDisposed)
    }

    @Test
    fun does_not_dispose_new_inner_source_disposable_WHEN_previous_inner_source_disposable_is_provided_after_new_one() {
        lateinit var innerObserver1: ObservableObserver<String?>
        val inner1 = observableUnsafe<String?> { innerObserver1 = it }
        lateinit var innerObserver2: ObservableObserver<String?>
        val inner2 = observableUnsafe<String?> { innerObserver2 = it }
        val innerDisposable2 = disposable()
        switchMapUpstreamAndSubscribe(listOf(inner1, inner2))

        source.onNext(0)
        source.onNext(1)
        innerObserver2.onSubscribe(innerDisposable2)
        innerObserver1.onSubscribe(disposable())

        assertFalse(innerDisposable2.isDisposed)
    }
```

Reply:
Done

Reply:
For some reason tests are failed for task ':reaktive:iosTest'.
`kotlin.UninitializedPropertyAccessException: lateinit property innerObserver2 has not been initialized` looks like inner source has not been subscribed at all. But jvm and js tests are passed.

Reply:
We should use AtomicReference instead of lateinit var:
```
val innerObserver1 = AtomicReference<ObservableObserver<String?>?>(null, true)
val inner1 = observableUnsafe<String?> { innerObserver1.value = it }
...
innerObserver1.value!!.onSubscribe(...)
```
Same for innerObserver2.